One-pass computation of line spacing

davidweichiang opened the pull request:
In #1488 the question came up of whether variable line spacing could be done in a single pass. This (sorry, rather large) patch does that. The score lines are printed in TeX with possible collisions, but with different parts of the score distinguished by different attributes, so that they can be put into the proper place in Lua. The advantages of this I see are:

- With two passes, the user might forget to run a second time, or be confused by large changes between the first and second pass.
- The two-pass computation was complicated, with control passing from TeX to Lua to TeX to Lua. It was difficult to inject the `\grechangedim`s at exactly the end of a line, which led to bugs.

The disadvantages are:

- `\grechangenextscorelinedim` and `\grechangenextscorelinecount` now only work for the distances and counts used for line spacing. I guess a helpful error should be printed if the user tries to change something else.
- It might seem conceptually strange to print a bad-looking score at first (which no one would see). In particular, the translation is printed right on top of the lyrics.

All tests pass except for:
gabc-output/glyphs/clef_change.gabc : this is #1633 I think
gabc-output/glyphs/nabc-edge-cases.gabc : this is #1634
gabc-output/bugs/fix-569.gabc : this is #1633
gabc-output/lines/3-lines.gabc : this is #1636
gabc-output/lines/2-lines.gabc : #1636 again
gabc-output/lines/5-lines.gabc : #1636 again
tex-output/dominican-flats/dominican-flats.tex : this one fails for me even in `ctan`, and the "correct" output looks wrong; I must have messed it up somehow
tex-output/bar-spacing/bar-spacing-new.tex : the difference is only in the lines used for debugging, and the new lines look more consistent to me?

Supersedes #1639.
Closes #1633, #1634, #1636.

davidweichiang committed:
One-pass computation of variable line heights

davidweichiang committed:
per-line spacing changes

davidweichiang committed:
remove code for two-pass line height computation

davidweichiang committed:
remove some dimens that are always 0pt or are redundant

davidweichiang committed:
remove a little bit more unused code

davidweichiang committed:
fix one failing test; trim more unused code

rpspringuel commented:
On the tests:

gabc-output/glyphs/clef_change.gabc : The new results look correct, while the old ones have a bad collision.
gabc-output/glyphs/nabc-edge-cases.gabc : I can confirm this fails, but the new results look fine to me.
gabc-output/bugs/fix-569.gabc : Again, new results seem to be better.
gabc-output/lines/3-lines.gabc : Definately a fix for the custos position.
gabc-output/lines/2-lines.gabc : Same as for 3-lines
gabc-output/lines/5-lines.gabc : Same as for 3-lines
tex-output/dominican-flats/dominican-flats.tex : This one doesn't fail for me.
tex-output/bar-spacing/bar-spacing-new.tex : I agree, the new results look like they are placing the debugging lines better.

tex-output/hidelyricsnotes/text-only.tex fails for me.  The document doesn't compile and the log indicates theirs a problem trying to do math that probably shouldn't be done in this case: `gregoriotex.lua:744: attempt to perform arithmetic on a nil value (field 'glyph_top')`

Some comments:
>With two passes, the user might forget to run a second time, or be confused by large changes between the first and second pass.

Multiple passes are a standard part of preparing TeX documents, so this would only be a factor for those who are new to the ecosystem.

That said, getting rid of requiring the second pass ourselves could be a huge time savings.  Loading GregorioTeX takes a substantial amount of time so while we probably can't get rid of the need for a second pass all by ourselves, if we can make it so that we are requiring the second pass, then we at least make the single-pass documents using Gregorio possible.  If we can solve some bugs at the same time, all the better.

> `\grechangenextscorelinedim` and `\grechangenextscorelinecount` now only work for the distances and counts used for line spacing. I guess a helpful error should be printed if the user tries to change something else.

The documentation indicates that `\grechangenextscorelinedim`is only supposed to work with `spaceabovelines`, `spacebeneathtext`, and `spacelinestext`.  That it worked with other distances is therefore, an undocumented feature, so your "restriction" would actually be an expansion of what it is documented to work with.

Besides, in both cases, I think this is what we originally intended both of these macros to be for.

>It might seem conceptually strange to print a bad-looking score at first (which no one would see). In particular, the translation is printed right on top of the lyrics.

This is hardly any more strange than the intermediate document produced in the middle of a multi-pass system.

rpspringuel commented:
Oh, and if you could create a parallel branch in the test repository with the new test results that would be appreciated (at least for the tests where we know the new result is because we're fixing the bugs).

davidweichiang commented:
> tex-output/dominican-flats/dominican-flats.tex : This one doesn't fail for me.

It must be because I'm working off of https://github.com/gregorio-project/gregorio-test/pull/360 and it's gotten out of sync.

> tex-output/hidelyricsnotes/text-only.tex fails for me. The document doesn't compile and the log indicates theirs a problem trying to do math that probably shouldn't be done in this case: gregoriotex.lua:744: attempt to perform arithmetic on a nil value (field 'glyph_top')

My fault, that one fails for me too, but it should work now with the last commit.

davidweichiang committed:
update docs/changelog

davidweichiang commented:
There's still a command `\greskipheightcomputation` that still does something, but not what it says. What should it be called instead?

davidweichiang committed:
remove \greskipheightcomputation

davidweichiang commented:
Since the speed difference due to `\greskipheightcomputation` seemed to be negligible (<0.1%), I took the liberty of removing it.

davidweichiang committed:
error message for invalid dims/counts for \grenextscorelinedim/count

rpspringuel commented:
I was away for the weekend and have been sick since I got back, so I haven’t had a chance to review anything posted since last Wednesday. I’m hoping to feel better and get back to work by Monday. 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

davidweichiang committed:
Apply additionaltopspacealt to commentary as well (fixes #1474).
Also fixes a bug: the commentary style was applied to the entire computation of the commentary raise, which was making the commentary appear too low.

davidweichiang commented:
With the last commit, three additional tests fail:
tex-output/commentary/commentary.tex: because the commentary height takes note height into account (#1474)
tex-output/PopulusSion/PopulusSion.tex: because a small bug was causing the commentary to be printed loo low
tex-output/per-line-spaces/per-line-spaces.tex: same as previous

davidweichiang commented:
> I was away for the weekend and have been sick since I got back, so I haven’t had a chance to review anything posted since last Wednesday. I’m hoping to feel better and get back to work by Monday.

No worries and hope you feel better soon!

rpspringuel commented on tex/gregoriotex-spaces.tex:
```diff
@@ -1590,6 +1590,9 @@
 % #1 : line number
 % #2-#4 : same as #1-#3 of \grechangedim
 \def\grechangenextscorelinedim#1#2#3#4{%
+  \IfStrEqCase{#2}{{abovelinestextheight}{}{noteadditionalspacelinestext}{}{spaceabovelines}{}{abovelinestextraise}{}{spacelinestext}{}{spacebeneathtext}{}{translationheight}{}}[%
```
Just to make sure I'm reading this right, you're using `\IfStrEqCase` to construct a `not (... and ... and ...)` sequence that will raise an error if this macro is called with an unsupported distance.

davidweichiang replied:
That's right.

davidweichiang replied:
Oh, actually, it's `if not (... or ... or ...) then print error`.

rpspringuel replied:
Yeah, that's what I meant.

rpspringuel commented on tex/gregoriotex-main.tex:
```diff
@@ -1117,10 +1067,10 @@
-% #3 is the low height
-% #4 is 1 if there is a translation somewhere
-% #5 is if 1 if we have space above the staff
+% #2 is the high height (currently ignored)
```
Can you foresee a future use for these arguments?  If not, I'd mark them as deprecated, not just ignored.  Once this PR is approved and merged, then we can work on the appropriate PR for the develop branch which will actually remove them (since they'd have to be removed from the executable, that's not a ctan eligible change).

davidweichiang replied:
I don't foresee a future use, no. How do I mark them as deprecated?

rpspringuel replied:
I've updated the Wiki page on the Deprecation process to cover this case: https://github.com/gregorio-project/gregorio/wiki/Deprecation-Process

rpspringuel commented:
I agree that the newly broken tests are all fixes to something which is currently wrong.

davidweichiang committed:
deprecation notices for unused arguments

davidweichiang commented:
Done, I think

rpspringuel commented on src/gregoriotex/gregoriotex-write.c:
```diff
@@ -4701,9 +4701,11 @@ void gregoriotex_write_score(FILE *const f, gregorio_score *const score,
     }
     fprintf(f, "\\GreBeginScore{%s}{%d}{%d}{%d}{%d}{%s}{%u}"
```
I'd put a deprecation comment here indicating what the string of arguments will look like after the deprecated ones are removed.

davidweichiang replied:
OK (also I wrote comments on two struct fields that can be removed).

davidweichiang committed:
more deprecation notices

davidweichiang committed:
a previous commit removed an unused argument to gregoriotex.at_score_beginning; this removes it from the call

davidweichiang committed:
silence error from sed

davidweichiang committed:
deprecate \GreInitialClefPosition

davidweichiang committed:
Because aa3c8ee6737b033a42c33c4cd56c2f99a834f06d shifted commentaries up a little bit, this shifts it back down a little bit. It's not possible to make it exactly like before (because the difference is font-dependent); instead, this places the commentary at the same height as the default for abovelinestext (30000sp*17 - 0.1cm).

davidweichiang commented:
I won't tinker with this anymore -- hopefully it's ready to go?

davidweichiang committed:
fix bug that could cause glyph heights to be carried over from one score to another

rpspringuel commented:
In reviewing this I've come across two unexplained test failures:

gabc-output/queuelength2_legacy.gabc
gabc-output/queuelength2.gabc

In both cases only the last three staves of the second page are effected.  The third to last stave is pushed down (though its lyrics are pushed up) and the last two staves are pushed up (by the same amount as the lyrics in the third to last stave, from the looks of it).

All the other test changes look good and are taken care of by the PR https://github.com/gregorio-project/gregorio-test/pull/373

davidweichiang commented:
Thanks for catching this -- I promised I wouldn't tinker any more and then I did! I believe this is a bug fix.

The first note of this score is below the staff (c=3), and I didn't fully look into why, but in `ctan` and in all commits of this PR up to the last one, the attribute `\gre@attr@glyph@bottom`=3 reappears in various places throughout the whole score. Ordinarily this seems to be harmless, but on line 14 (third from last), and again I didn't fully look into why, the 3 actually appears as an attribute on an hbox that matters. The line spacing algorithm thinks this line has a `c` note on it, and the lyrics are typeset too low.

<img width="68" alt="image" src="https://github.com/user-attachments/assets/021ba310-00ee-4386-92e6-8b2d68c4cca9" />

I believe the shift in the last two staves is because in the old code, line 14's height was greater than `\baselineskip`, so `\lineskip` (=1pt, I think) was inserted.

In the last commit, `\gre@attr@glyph@top` and `\gre@attr@glyph@bottom` are reset after each glyph, so the problem goes away.

<img width="68" alt="image" src="https://github.com/user-attachments/assets/4edbc079-ad4c-499d-ba74-f328bc1a1396" />

rpspringuel commented:
That explains why the lyrics (and the staves below) move up, but why does the 3rd to last set of staff lines move down?  There doesn't appear to be anything on the 4th to last line that would require extra space below that stave nor anything on the 3rd to last line that would require extra space above that line.

davidweichiang commented:
That's because the baseline of each staff is the baseline of the lyrics (or translation if any), and TeX tries to ensure that the distance between baselines is `\baselineskip`. So lyrics moving up really means staff moving down.

(I think a case could be made for the baseline to instead be the lowest staff line, but that's a separate issue!)

rpspringuel commented:
This is why I find `\baselineskip` so confusing.

Okay, so it looks like everything this good at this point, just need to have those altered tests accepted into the test repository.

davidweichiang commented:
Done, thanks!
